### PR TITLE
Integrate ImagingDataChunkIterator with write tools

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
## Motivation

PR has two linked parts needed for my current conversion.

1a) Adds the new `ImagingDataChunkIterator` to the `add_two_photon` helper function in `tools.roiextractors`.
1b) TODO - need to add to `write_imaging` as well.
2) 'technically' adds support to the `ImagingDataChunkIterator` for volumetric data (i.e., an extra axis on the end for depth). Works on top of existing roiextractors API with no actual volumetric support needed (just depends on image size retrieved through the extractor)

- [x] Also TODO is remove the print statements.

## How to test the behavior?
@weiglszonja What tests would you want for this? Could you add them if you have the time?

## Checklist

- [x] Have you thoroughly read our [Developer Guide](https://neuroconv.readthedocs.io/en/main/developer_guide.html)?
- [x] Have you ensured the PR description clearly describes the problem and solutions?
- [x] Have you checked to ensure that there aren't other open or previously closed [Pull Requests](https://github.com/catalystneuro/neuroconv/pulls) for the same change?
- [x] If this PR fixes an issue, is the first line of the PR description `fix #XXX` where `XXX` is the issue number?
- [x] Did you update the [CHANGLOG.md](https://github.com/catalystneuro/neuroconv/blob/main/CHANGELOG.md) file on your branch to describe your changes?

### CODE REVIEW

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 1]
```diff
@@ -301,16 +316,26 @@ def add_two_photon_series(
...
+            for i in range(imaging.get_num_frames()):
+                yield imaging.get_frames(frame_idxs=[i]).squeeze().T
+
+        if "buffer_size" not in iterator_options:
+            iterator_options.update(buffer_size=10)
+        data = H5DataIO(
+            data=DataChunkIterator(data_generator(imaging), **iterator_options),
+            compression=True,
+        )
+        two_p_series_kwargs.update(data=data)
```
Comment (h-mayorquin) [inline src/neuroconv/tools/roiextractors/roiextractors.py:338]: We should wrap this in an aux function.
Adding the details of which iterator is wrapped and how the parameters are propagated are details that do not concern the high level user of the `add_two_photon_series` (as the only things that matters is that the data is wrapped in an `h5dataio` iterator). Those details can easily be abstracted away to make this more readable.

**Review by h-mayorquin — CHANGES_REQUESTED** [review]

File: src/neuroconv/tools/roiextractors/roiextractors.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/roiextractors/roiextractors.py:338]: IDK how to do that - can you make a splinter PR pointing to this one as the base to do what you're imagining?

Comment (h-mayorquin) [reply src/neuroconv/tools/roiextractors/roiextractors.py:338]: I mean something like what we did for the ecephys pipeline:

https://github.com/catalystneuro/neuroconv/blob/27357f8145746f2279e580b8a5565eaa26a98f96/src/neuroconv/tools/spikeinterface/spikeinterface.py#L491-L550

**Comment by h-mayorquin** [comment]
Also, check out the tests that I did recently for the ecephys side. Maybe I missed something over there as well:

https://github.com/catalystneuro/neuroconv/blob/27357f8145746f2279e580b8a5565eaa26a98f96/tests/test_internals/test_si.py#L358

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 2]
```diff
@@ -272,16 +273,30 @@ def add_image_segmentation(nwbfile: NWBFile, metadata: dict) -> NWBFile:
...
 
 def add_two_photon_series(
-    imaging, nwbfile, metadata, buffer_size=10, use_times=False, two_photon_series_index: int = 0
+    imaging,
+    nwbfile,
+    metadata,
+    two_photon_series_index: int = 0,
+    use_times=False,
+    buffer_size: Optional[int] = None,
+    iterator_type: Optional[str] = None,
```
Comment (weiglszonja) [inline src/neuroconv/tools/roiextractors/roiextractors.py:282]:
```suggestion
    iterator_type: str = "v2",
```
I'd use the new one by default, which can be modified if someone insists to use the other one.

File: src/neuroconv/tools/roiextractors/roiextractors.py (hunk 1 above)
Comment (weiglszonja) [reply src/neuroconv/tools/roiextractors/roiextractors.py:338]: thanks @h-mayorquin for sharing this, I tried to do something similar here: https://github.com/catalystneuro/neuroconv/pull/91

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 3]
```diff
@@ -272,16 +273,30 @@ def add_image_segmentation(nwbfile: NWBFile, metadata: dict) -> NWBFile:
...
     """
-
     if use_times:
         warn("Keyword argument 'use_times' is deprecated and will be removed on or after August 1st, 2022.")
+    if buffer_size:
+        warn(
+            "Keyword argument 'buffer_size' is deprecated and will be removed on or after September 1st, 2022."
+            "Specify as a key in the new 'iterator_options' dictionary instead."
+        )
+    iterator_type = iterator_type or "v2"
```
Comment (weiglszonja) [inline src/neuroconv/tools/roiextractors/roiextractors.py:297]:
```suggestion

```
If we use "v2" as default this line is not needed

Comment (h-mayorquin) [reply src/neuroconv/tools/roiextractors/roiextractors.py:297]: This is the case for `add_electrical_series`. `v2` is also the default.

File: src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py [hunk 4]
```diff
@@ -77,11 +77,12 @@ def run_conversion(
         metadata: Optional[dict] = None,
         overwrite: bool = False,
         stub_test: bool = False,
+        stub_frames: int = 100,
         save_path: OptionalFilePathType = None,
     ):
 
         if stub_test:
-            stub_frames = min([100, self.imaging_extractor.get_num_frames()])
+            stub_frames = min([stub_frames, self.imaging_extractor.get_num_frames()])
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py:85]: I needed this one for my current conversion since even 5 frames is almost half a GB of data. This kind of thing is really only needed for Imaging of volumetric data, ecephys interfaces should have no trouble similar to this for a very long time.

That said, it is perhaps tangential to the rest of the PR so I could split it into a separate PR if requested

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 5]
```diff
@@ -331,6 +343,54 @@ def data_generator(imaging):
...
+
+    def data_generator(imaging):
+        for i in range(imaging.get_num_frames()):
+            yield imaging.get_frames(frame_idxs=[i]).squeeze().T
+
+    assert iterator_type in ["v1", "v2", None], "'iterator_type' must be either 'v1', 'v2' (recommended), or None."
+    iterator_options = dict() if iterator_options is None else iterator_options
+
+    if iterator_type is None:
+        return
```
Comment (weiglszonja) [inline src/neuroconv/tools/roiextractors/roiextractors.py:384]: @CodyCBakerPhD 
If this returns None, then this will cause an error downstream in `hdmf` that we're trying to add data which is None. Should this be handled in `add_two_photon_series`?

Comment (CodyCBakerPhD) [reply src/neuroconv/tools/roiextractors/roiextractors.py:384]: Oh, whoops, I meant to do something here lol

File: src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py (hunk 4 above)
Comment (weiglszonja) [reply src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py:85]: This is good to have here, I'm just thinking whether it would make sense to have `stub_options` as dict instead of just the `stub_frames`, but I guess it can be changed later if we wanted to stub by image size later?

Comment (CodyCBakerPhD) [reply src/neuroconv/datainterfaces/ophys/baseimagingextractorinterface.py:85]: `stub_options` isn't a bad idea in the long run, as we could include the `stub_test: bool` functionality within it, such as `enable: bool`. Would also be useful for channel-wise stubbing in ecephys for the previous observation interval argument.

That's probably its own entire thing though... especially as we'd have to start a definition for the sub-schema of that argument. Maybe best left for subsequent PR?

File: src/neuroconv/tools/roiextractors/roiextractors.py (hunk 5 above)
Comment (weiglszonja) [reply src/neuroconv/tools/roiextractors/roiextractors.py]: Great, looks good to me.

**Review by weiglszonja — APPROVED** [review]

File: src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py [hunk 6]
```diff
@@ -97,19 +96,25 @@ def _get_scaled_buffer_shape(self, buffer_gb: float, chunk_shape: tuple) -> tupl
...
 
     def _get_data(self, selection: Tuple[slice]) -> np.ndarray:
         data = self.imaging_extractor.get_video(
             start_frame=selection[0].start,
             end_frame=selection[0].stop,
         )
-
-        return data.transpose((0, 2, 1))[(slice(0, self.buffer_shape[0]),) + selection[1:]]
+        tranpose_axes = (0, 2, 1) if len(data.shape) == 3 else (0, 2, 1, 3)
+        return data.transpose(tranpose_axes)[(slice(0, self.buffer_shape[0]),) + selection[1:]]
```
Comment (h-mayorquin) [inline src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py:120]: Out of curiosity, why the `selection[1:]` here?

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 7]
```diff
@@ -272,16 +265,29 @@ def add_image_segmentation(nwbfile: NWBFile, metadata: dict) -> NWBFile:
 
 
 def add_two_photon_series(
-    imaging, nwbfile, metadata, buffer_size=10, use_times=False, two_photon_series_index: int = 0
+    imaging,
```
Comment (h-mayorquin) [inline src/neuroconv/tools/roiextractors/roiextractors.py:268]: Would be nice if we could add annotation types to these now that we are a it.

**Review by h-mayorquin — APPROVED** [review]
Everything looks good to me.

File: src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py (hunk 6 above)
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py:120]: `selection[0]` is the main slice through `get_video` until ROIExtractors supports efficient image slicing (if it even can for certain extractors... safer to just assume frame-wise access)

Then the remaining axes of selection could be 2 or 3 (possibly 5 for sbx?) dimensional, but that's handled automatically by the parent class.

File: src/neuroconv/tools/roiextractors/roiextractors.py (hunk 7 above)
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/roiextractors/roiextractors.py:268]: Sure I'll add those after in an easy formatting/polishing PR

File: src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py (hunk 6 above)
Comment (h-mayorquin) [reply src/neuroconv/tools/roiextractors/imagingextractordatachunkiterator.py:120]: Thanks.
